internal/pkg/cli/app_list.go

```diff
@@ -0,0 +1,209 @@
+	paddingChar            = ' ' // character in between columns.
+	noAdditionalFormatting = 0
+)
+
+// ListAppOpts contains the fields to collect for listing an application.
+type ListAppOpts struct {
+	ShouldOutputJSON    bool
+	ShouldShowLocalApps bool
+
+	applications  []*archer.Application
+	appLister     archer.ApplicationLister
+	projectGetter archer.ProjectGetter
+	projectLister archer.ProjectLister
+
+	ws archer.Workspace
+	w  io.Writer
+
+	*GlobalOpts
+}
+
+func (opts *ListAppOpts) selectProject() (string, error) {
+	projs, err := opts.projectLister.ListProjects()
+	if err != nil {
+		return "", err
+	}
+	var projNames []string
+	for _, proj := range projs {
+		projNames = append(projNames, proj.Name)
+	}
+	if len(projNames) == 0 {
+		log.Infoln("There are no projects to select.")
+		return "", nil
+	}
+	proj, err := opts.prompt.SelectOne(
+		applicationListProjectNamePrompt,
+		applicationListProjectNameHelpPrompt,
+		projNames,
+	)
+	return proj, err
+}
+
+func (opts *ListAppOpts) localAppsFilter(appNames []string) {
+	var filtered []*archer.Application
+	isLocal := make(map[string]bool)
+	for _, name := range appNames {
+		isLocal[name] = true
+	}
+	for _, app := range opts.applications {
+		if _, ok := isLocal[app.Name]; !ok {
+			continue
+		}
+		filtered = append(filtered, app)
+	}
+	opts.applications = filtered
+}
+
+// Ask asks for fields that are required but not passed in.
+func (opts *ListAppOpts) Ask() error {
+	if opts.ProjectName() != "" {
+		return nil
+	}
+	projectName, err := opts.selectProject()
+	if err != nil {
+		return fmt.Errorf("failed to get project name: %w", err)
+	}
+	opts.projectName = projectName
+
+	return nil
+}
+
+// Execute lists the applications through the prompt.
+func (opts *ListAppOpts) Execute() error {
+	// Ensure the project actually exists before we try to list its applications.
+	if _, err := opts.projectGetter.GetProject(opts.ProjectName()); err != nil {
+		return err
+	}
+
+	apps, err := opts.appLister.ListApplications(opts.ProjectName())
+	if err != nil {
+		return err
+	}
+	opts.applications = apps
+
+	if opts.ShouldShowLocalApps {
+		localAppManifests, err := opts.ws.Apps()
+		if err != nil {
+			return fmt.Errorf("failed to get local app manifests: %w", err)
+		}
+		var localAppNames []string
+		for _, appManifest := range localAppManifests {
+			localAppNames = append(localAppNames, appManifest.AppName())
+		}
+		opts.localAppsFilter(localAppNames)
+	}
+
+	var out string
+	if opts.ShouldOutputJSON {
+		data, err := opts.jsonOutput()
+		if err != nil {
+			return err
+		}
+		out = data
+		fmt.Fprintf(opts.w, out)
+	} else {
+		opts.humanOutput()
+	}
+
+	return nil
+}
+
+func (opts *ListAppOpts) humanOutput() {
+	writer := tabwriter.NewWriter(opts.w, minCellWidth, tabWidth, cellPaddingWidth, paddingChar, noAdditionalFormatting)
+	fmt.Fprintf(writer, "%s\t%s\n", "Name", "Type")
+	nameLengthMax := 0
+	typeLengthMax := 0
```

Comment:
maybe that should start with len("Name") and len("Type")?

Reply:
:+1: you are so good!